volestipy/volestipy/volestipy.pyx

```diff
@@ -3,26 +3,146 @@
 #cython: boundscheck=False
 #cython: wraparound=False
 
+# Global dependencies
 import os
 import sys
 import numpy as np
 cimport numpy as np
 from libcpp cimport bool
 from cpython cimport bool
 
-# set the time
+# For the preprocess step, we need the following dependencies
+import scipy . sparse as sp
+import gurobipy as gp
+from gurobipy import GRB
+
+# ----------------------------------------------------------------------------------
+
+# Set the time
 def get_time_seed():
    import random
    import time
    return int(time.time())
 
+# Build a Python function to pre-process the metabolic network; meaning to remove really "small" facets.
+# This function will be implemented by making use of the Gurobi solver
+def pre_process(A, b, Aeq, beq):
+
+   d = A.shape[1] ; m = Aeq.shape[0] ; n = Aeq.shape[1]
+   Aeq_new = Aeq
+   beq_new = beq
+   A_new = np.zeros((0,d))
+   b_new = []    # this need to be a vector; we do not know its length
+
+   try:
+
+      # Create a new model
+      model = gp.Model("preProcHPol")
+
+      # Create variables
+      x = model.addMVar(shape = d, vtype = GRB.CONTINUOUS , name = "x", lb = -GRB.INFINITY, ub = GRB.INFINITY)
+
+      # Make sparse Aeq
+      Aeq_sparse = sp.csr_matrix(Aeq)
+
+      # Make A sparse
+      A_sparse = sp.csr_matrix(A)
+
+      # Set the b and beq vectors as numpy vectors
+      b = np.array(b)
+      beq = np.array(beq)
+
+      # Add constraints
+      model.addMConstrs(Aeq_sparse, x, '=', beq, name = "c")
+
+      # Update the model to include the constraints added
+      model.update()
+
+      #######################
+      # After getting the constraints you need to add the bounds; ObjBound might work: https://www.gurobi.com/documentation/9.0/refman/objbound.html#attr:ObjBound
+      # to start with, avoid ObjBound and do that the same way as Aeq but with unequalities this time
+      #######################
+
+      # Add constraints for the uneqalities of A
+      model.addMConstrs(A_sparse, x, '<', b, name = "d")
+
+      # Update the model with the extra constraints and then print it
+      model.update()
+      model.display()
+
+      # Loop through the lines of the A matrix, set objective function for each and run the model
+      for i in range(A.shape[0]):
+
+         # Set the ith row of the A matrix as the objective function
+         objective_function = A[i,]
+
+         # Set the objective function in the model
+         model.setMObjective(None, objective_function, 0.0, None, None, x, GRB.MINIMIZE)
+         model.update()
+
+         # Optimize model
+         model.optimize ()
+
+         # If optimized, print the solution
+         status = model.status
+         if status == GRB.OPTIMAL:
+            solution = model.getAttr('x')
+            print("\n\n***The solution for the MAX case is:")
+            print(solution)
```

Comment:
Do we need printing here? I think printing should be used only in tests

Reply:
Nope. I removed the ```solution``` variable completely